OpenRA.Game/Traits/Target.cs

```diff
@@ -173,6 +163,48 @@ public bool IsInRange(WPos origin, WDist range)
 			return Positions.Any(t => (t - origin).HorizontalLengthSquared <= range.LengthSquared);
 		}
 
+		public bool IsOwnerTargetable(Actor self)
+		{
+			if (this.FrozenActor != null)
+				return this.FrozenActor.Visible;
+			else if (this.Actor != null)
+				return self.Owner.CanTargetActor(this.Actor);
+
+			return true;
+		}
+
+		public Target TryUpdateFrozenActorTarget(Actor self)
+		{
+			// If target was a frozen actor got replaced with an actor, try switching
+			if (this.FrozenActor != null && this.FrozenActor.Actor != null)
+			{
+				var loc = self.World.Map.CellContaining(this.FrozenActor.CenterPosition);
+				foreach (var actor in self.World.ActorMap.GetActorsAt(loc))
+				{
+					if (actor.ActorID == this.FrozenActor.ID && self.Owner.CanTargetActor(actor))
```

Comment:
`FrozenActor.ID` is by *definition* `FrozenActor.Actor.ID`, so the `GetActorsAt` query is unnecessary.  Just use `FrozenActor.Actor` directly.